output/cloud/expv2/hdr.go

```diff
@@ -0,0 +1,273 @@
+	// LastNotZeroBucket represents the index of the last bucket
+	// with a significant counter in the Buckets slice (a not zero value).
+	// In this way, all the buckets after can be omitted.
+	LastNotZeroBucket uint32
+
+	// Max is the absolute maximum observed value.
+	Max float64
+
+	// Min is the absolute minimum observed value.
+	Min float64
+
+	// Sum is the sum of all observed values.
+	Sum float64
+
+	// Count is counts the amount of observed values.
+	Count uint32
+}
+
+// newHistogram creates an histogram of the provided values.
+//
+// TODO: it isn't really useful so we may consider to drop it.
+func newHistogram() histogram {
+	return histogram{}
+}
+
+// addToBucket increments the counter of the bucket of the provided value.
+// If the value is lower or higher than the trackable limits
+// then it is counted into specific buckets. All the stats are also updated accordingly.
+func (h *histogram) addToBucket(v float64) {
+	if h.Count == 0 {
+		h.Max, h.Min = v, v
+	} else {
+		if v > h.Max {
+			h.Max = v
+		}
+		if v < h.Min {
+			h.Min = v
+		}
+	}
+
+	h.Count++
+	h.Sum += v
+
+	if v > highestTrackable {
+		h.ExtraHighBucket++
+		return
+	}
+	if v < lowestTrackable {
+		h.ExtraLowBucket++
+		return
+	}
+
+	index := resolveBucketIndex(v)
+
+	blen := uint32(len(h.Buckets))
+	if blen == 0 {
+		h.FirstNotZeroBucket = index
+		h.LastNotZeroBucket = index
+		h.Buckets = make([]uint32, 1, 32)
+	} else {
+		if index < h.FirstNotZeroBucket {
+			h.growLeft(index)
+			h.FirstNotZeroBucket = index
+		}
+		if index > h.LastNotZeroBucket {
+			h.growRight(index)
+			h.LastNotZeroBucket = index
+		}
+	}
+
+	h.Buckets[index-h.FirstNotZeroBucket]++
+}
+
+func (h *histogram) growLeft(index uint32) {
+	if h.FirstNotZeroBucket <= index {
+		panic("buckets is already contains the requested index")
+	}
+
+	newLen := (h.FirstNotZeroBucket - index) + uint32(len(h.Buckets))
+
+	// there is enough capacity in the undrelying array
+	// reuse it slicing without allocate a new array
+	if uint32(cap(h.Buckets)) >= newLen {
+		upIndex := len(h.Buckets) - 1
+		h.Buckets = h.Buckets[:newLen]
+		h.Buckets = append(h.Buckets[upIndex+1:], h.Buckets[:upIndex+1]...)
+		return
+	}
+
+	// there isn't enough capacity, allocate a new slice
+	// able to contain the new size
+
+	newBuckets := make([]uint32, newLen)
+	copy(newBuckets[h.FirstNotZeroBucket-index:], h.Buckets)
+	h.Buckets = newBuckets
+}
+
+// growRight expands the buckets slice
+// with zeros up to the required index.
+// If it array has enough capacity then it reuses it without allocate.
+func (h *histogram) growRight(index uint32) {
+	if h.LastNotZeroBucket >= index {
+		panic("buckets is already bigger than requested index")
+	}
+
+	newLen := index - h.FirstNotZeroBucket + 1
+
+	if uint32(cap(h.Buckets)) > newLen {
+		// See https://go.dev/ref/spec#Slice_expressions
+		// "For slices, the upper index bound is
+		// the slice capacity cap(a) rather than the length"
+		h.Buckets = h.Buckets[:newLen]
+	} else {
```

Comment:
I don't see how this slice will ever have different cap and len as you never use `append` with it in a way that will introduce that.

So IMO these checks(and the one in growLeft) are pointless - you will always need to allocate a new one as you never actually let go allocate more for you.

Reply:
[At the begin](https://github.com/grafana/k6/blob/128e55a042c258d238786b9ba887093884299316/output/cloud/expv2/hdr.go#L120-L123) we allocate an empty slice with some cap (my expectation is: it saves from all that aggregation where we have very close/similar values). It should be most of the case for successful tests.

Reply:
I did miss that. Why don't we create this in the constructor? 

I would argue that a lot of the code will be much more clean if creating a new histogram takes the first value. This way a bunch of the `if`s just dissappaer. I guess there the usage will be slightly different but arguably not that much :thinking:

Reply:
Take a look at https://github.com/grafana/k6/pull/3027/commits/e01da6c2ca4522dd0041c68b1f187943e8b2385d, is it better? Note that it allocates without being useful in the case of un trackable values. But probably we don't care considering it shouldn't be a high percentage of the cases.